File: CobongMemo/app/src/main/java/cobong/jeongwoojin/cobongmemo/cobongmemo/view/memo/textmemo/TextMemoViewModel.java
package cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.textmemo;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import cobong.jeongwoojin.cobongmemo.cobongmemo.model.MemoListItem;
import cobong.jeongwoojin.cobongmemo.cobongmemo.model.MemoRepository;

public class TextMemoViewModel extends AndroidViewModel {

    private MutableLiveData<MemoListItem> item = new MutableLiveData<>();
    private TextMemoNavigator navigator;

    public TextMemoViewModel(Application application) {
        super(application);
    }

    //메모 수정
    public void onEditClick() {
        navigator.onEditClick();
    }

    //메모 삭제
    public void onDeleteClick() {
        navigator.onDeleteClick();
    }

    //뒤로가기
    public void onExitClick() {
        navigator.onExitClick();
    }

    //작성완료
    public void onWriteClick() {
        navigator.onWriteClick();
    }


    public TextMemoNavigator getNavigator() {
        return navigator;
    }

    public void setNavigator(TextMemoNavigator navigator) {
        this.navigator = navigator;
    }

    public MutableLiveData<MemoListItem> getItem() {
        return item;
    }

    public void setItem(MemoListItem item) {
        this.item.setValue(item);
    }


    public void insertTextMemo(String title, String subTitle, String content) {
        MemoRepository.Companion.getInstance(getApplication()).insertTextMemo(title, subTitle, content);
    }

    public void updateTextMemo(int index, String title, String subTitle, String content) {
        MemoRepository.Companion.getInstance(getApplication()).updateTextMemo(index, title, subTitle, content);
    }

    public void deleteTextMemo() {
        MemoRepository.Companion.getInstance(getApplication()).deleteTextMemo(item.getValue().getIndex());
    }

}

File: CobongMemo/app/src/main/java/cobong/jeongwoojin/cobongmemo/cobongmemo/view/memo/textmemo/TextMemoViewModel.kt
package cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.textmemo

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import cobong.jeongwoojin.cobongmemo.cobongmemo.MemoApplication
import cobong.jeongwoojin.cobongmemo.cobongmemo.common.Event
import cobong.jeongwoojin.cobongmemo.cobongmemo.model.memo.MemoItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class TextMemoViewModel(application: Application) : AndroidViewModel(application) {

    var item = MutableLiveData<MemoItem>()

    //메모 수정
    private val _editMemoEvent = MutableLiveData<Event<Unit>>()
    val editMemoEvent: LiveData<Event<Unit>> = _editMemoEvent

    //메모 삭제
    private val _deleteMemoEvent = MutableLiveData<Event<Unit>>()
    val deleteMemoEvent: LiveData<Event<Unit>> = _deleteMemoEvent

    //뒤로가기
    private val _exitClickEvent = MutableLiveData<Event<Unit>>()
    val exitClickEvent: LiveData<Event<Unit>> = _exitClickEvent

    //작성 완료
    private val _writeClickEvent = MutableLiveData<Event<Unit>>()
    val writeClickEvent: LiveData<Event<Unit>> = _writeClickEvent

    //메모 수정
    fun onEditClick() {
        _editMemoEvent.value = Event(Unit)
    }

    //메모 삭제
    fun onDeleteClick() {
        _deleteMemoEvent.value = Event(Unit)
    }

    //뒤로가기
    fun onExitClick() {
        _exitClickEvent.value = Event(Unit)
    }

    //작성완료
    fun onWriteClick() {
        _writeClickEvent.value = Event(Unit)
    }

    fun insertTextMemoByRoom(title: String, subTitle: String, content: String) =
        viewModelScope.launch (Dispatchers.IO){
            MemoApplication.memoRepository.insertByRoom(
                MemoItem(
                    index = null,
                    title = title,
                    subTitle = subTitle,
                    memoType = "text",
                    content = content,
                    voiceId = null,
                    handwriteId = null
                )
            )
        }

    fun updateTextMemoByRoom(title: String, subTitle: String, content: String) =
        viewModelScope.launch (Dispatchers.IO){
            MemoApplication.memoRepository.updateByRoom(
                MemoItem(
                    index = item.value?.index,
                    title = title,
                    subTitle = subTitle,
                    memoType = item.value?.memoType,
                    content = content,
                    voiceId = item.value?.voiceId,
                    handwriteId = item.value?.handwriteId
                )
            )
        }

    fun deleteTextMemoByRoom() =
        viewModelScope.launch (Dispatchers.IO){
            MemoApplication.memoRepository.deleteMemoNullableByRoom(
               item.value
            )
        }

}
